### PR TITLE
Cleanup reflected regions finder

### PR DESCRIPTION
**Pull request by adonath**
**Description**

This pull request partly addresses #2267 and cleans up the implementation of the `ReflectedRegionsFinder`. the most relevant change here is to replace the `.setup()` methods by a `lazyproperty` pattern. This allowed me to replace the derived quantities from the finder configuration by lazy properties. I think the resulting implementation is very simple and clean.

### CODE REVIEW

File: gammapy/makers/background/reflected.py [hunk 1]
```diff
@@ -179,74 +159,69 @@ def _region_angular_size(pixels, center):
...
-        # Extract all pixcoords in the geom
-        X, Y = geom.get_pix()
-        ONpixels = PixCoord(X[mask], Y[mask])
+    @lazyproperty
+    def exclusion_mask_ref(self):
+        """Exclusion mask reprojected"""
+        if self.exclusion_mask:
+            mask = self.exclusion_mask.interp_to_geom(self.geom_ref, fill_value=True)
+        else:
+            mask = WcsNDMap.from_geom(geom=self.geom_ref, data=1)
```
Comment (registerrier) [inline gammapy/makers/background/reflected.py:168]: why not creating directly the Map as boolean?

File: gammapy/makers/background/reflected.py [hunk 2]
```diff
@@ -179,74 +159,69 @@ def _region_angular_size(pixels, center):
...
+        return mask
 
+    @lazyproperty
+    def excluded_pix_coords(self):
+        """Excluded pix coords"""
         # find excluded PixCoords
-        mask = self.reference_map.data == 0
-        self.excluded_pixcoords = PixCoord(X[mask], Y[mask])
+        pix_y, pix_x = np.where(~self.exclusion_mask_ref.data)
+        return PixCoord(pix_x, pix_y)
```
Comment (registerrier) [inline gammapy/makers/background/reflected.py:177]: Is it always equivalent to X[]?

File: gammapy/makers/background/reflected.py [hunk 3]
```diff
@@ -179,74 +159,69 @@ def _region_angular_size(pixels, center):
...
+    def run(self):
+        """Find reflected regions.
 
-        self.reflected_regions = reflected_regions
+        Returns
+        -------
+        regions : list of `SkyRegion`
+            Reflected regions
+        """
+        self.reset_cache()
```
Comment (registerrier) [inline gammapy/makers/background/reflected.py:205]: Does it need to reset the cache since the `RegionsFinder` is re-instantiated at each `run` of the `ReflectedRegionsBackgroundMaker`?
  You could also define getters and no setters for the `center` and `region` to force users to create a new `RegionFinder` for a new set of region and center position.

File: gammapy/makers/background/tests/test_reflected.py [hunk 4]
```diff
@@ -114,19 +110,18 @@ def test_find_reflected_regions(
...
 center = SkyCoord(0.5, 0.0, unit="deg")
 other_region_finder_param = [
     (RectangleSkyRegion(center, 0.5 * u.deg, 0.5 * u.deg, angle=0 * u.deg), 3),
     (RectangleSkyRegion(center, 0.5 * u.deg, 1 * u.deg, angle=0 * u.deg), 1),
-    (RectangleSkyRegion(center, 0.5 * u.deg, 1 * u.deg, angle=90 * u.deg), 0),
+    (RectangleSkyRegion(center, 0.5 * u.deg, 1 * u.deg, angle=90 * u.deg), 1),
     (EllipseSkyRegion(center, 0.1 * u.deg, 1 * u.deg, angle=0 * u.deg), 2),
     (EllipseSkyRegion(center, 0.1 * u.deg, 1 * u.deg, angle=60 * u.deg), 3),
-    (EllipseSkyRegion(center, 0.1 * u.deg, 1 * u.deg, angle=90 * u.deg), 0),
+    (EllipseSkyRegion(center, 0.1 * u.deg, 1 * u.deg, angle=90 * u.deg), 2),
```
Comment (registerrier) [inline gammapy/makers/background/tests/test_reflected.py:124]: Why the change here?

**Review by registerrier — DISMISSED** [review]
Thanks @adonath . It seems indeed clearer with a pattern where everything is defined on `__init__`  and `run` returns the list of regions.
I would even enforce that `center` and `region` cannot be set to remove the need for a clear_cache.

There is conflict with the new version of `Analysis` introduced in #3424

File: gammapy/makers/background/reflected.py (hunk 1 above)
Comment (adonath) [reply gammapy/makers/background/reflected.py:168]: Good point, I tried:
```python3
m = Map.from_geom(geom=geom, data=True)
m = Map.from_geom(geom=geom, data=1, dtype=bool)
```
Neither seems to work, I'll investigate why...

File: gammapy/makers/background/reflected.py (hunk 2 above)
Comment (adonath) [reply gammapy/makers/background/reflected.py:177]: Yes, I think so...I haven't checked the performance, but using `np.where()` seemed simpler to me...

File: gammapy/makers/background/reflected.py (hunk 3 above)
Comment (adonath) [reply gammapy/makers/background/reflected.py:205]: Good point, indeed if `center` and `region` cannot be changed we can probably remove the reset cache pattern. I think in general the pattern is not bad, but I agree further restricting the attributes is even better...

File: gammapy/makers/background/tests/test_reflected.py (hunk 4 above)
Comment (adonath) [reply gammapy/makers/background/tests/test_reflected.py:124]: I will try to understand the change and comment again...

File: gammapy/makers/background/reflected.py (hunk 1 above)
Comment (adonath) [reply gammapy/makers/background/reflected.py:168]: Done

File: gammapy/makers/background/reflected.py (hunk 3 above)
Comment (adonath) [reply gammapy/makers/background/reflected.py:205]: Looking more closely, it would require to declare all the attributes as non-settable attributes, which creates a lot of extra code. I the only way to further improve and simplify would be to move more code to `RegionGeom` as it already bundles the `region` and `wcs`. This way one could cache on the `RegionGeom` object.

File: gammapy/makers/background/tests/test_reflected.py (hunk 4 above)
Comment (adonath) [reply gammapy/makers/background/tests/test_reflected.py:124]: # Before:
Screenshot 2021-07-14 at 10 14 50
# After:
Screenshot 2021-07-14 at 10 12 35

I think placing the regions for the last column is within the numerical precision of the algorithm, which depends on the bin  size of the tangential WCS. I'm not 100% why the behaviour changed, but result still seems OK.

Comment (adonath) [reply gammapy/makers/background/tests/test_reflected.py:124]: In general one could change to a less sensitive test, but maybe it's better to keep the more sensitive one and check the results once by hand.
